lib/gc/Transforms/IterativeTilingAndFusion.cpp

```diff
@@ -0,0 +1,853 @@
+          getConstantIntValue(iterDomain[iterPosition].size);
+      FailureOr<int64_t> cstTileSizes =
+          ValueBoundsConstraintSet::computeConstantBound(
+              presburger::BoundType::UB, tileSizes[resultExpr.index()], nullptr,
+              true);
+      if (!cstIterDomain || failed(cstTileSizes) ||
+          cstIterDomain != cstTileSizes)
+        return failure();
+    }
+  }
+  return success();
+}
+
+static LogicalResult
+exactTilingOnPackUnPackFilter(RewriterBase &rewriter,
+                              OffsetSizeAndStrideOpInterface candidate,
+                              CandidateDefOrUse defOrUse) {
+  if (!isa<tensor::PackOp, tensor::UnPackOp>(defOrUse.ownerOp))
+    return success();
+
+  SmallVector<OpFoldResult> tileSizes = candidate.getMixedSizes();
+  // collect target TileSizes and InnerTileSize to compare
+  SmallVector<OpFoldResult> targetTileSizes, targetInnerTileSizes;
+  if (auto packOp = dyn_cast<tensor::PackOp>(defOrUse.ownerOp)) {
+    // tileSize comes from OpResult
+    if (defOrUse.isDef()) {
+      targetInnerTileSizes = packOp.getInnerTiles();
+      targetTileSizes = llvm::to_vector(
+          ArrayRef(tileSizes).take_back(targetInnerTileSizes.size()));
+    } else {
+      // tileSize comes from OpOperand
+      targetTileSizes = llvm::to_vector(tileSizes);
+      DenseMap<int64_t, OpFoldResult> dimAndTileMapping =
+          packOp.getDimAndTileMapping();
+      targetInnerTileSizes.resize(dimAndTileMapping.size());
+      for (const auto &dimAndTile : dimAndTileMapping) {
+        targetInnerTileSizes[dimAndTile.first] = dimAndTile.second;
+      }
+    }
+  } else if (auto unPackOp = dyn_cast<tensor::UnPackOp>(defOrUse.ownerOp)) {
+    // tileSize comes from OpResult
+    if (defOrUse.isDef()) {
+      targetTileSizes = llvm::to_vector(tileSizes);
+      DenseMap<int64_t, OpFoldResult> dimAndTileMapping =
+          unPackOp.getDimAndTileMapping();
+      targetInnerTileSizes.resize(dimAndTileMapping.size());
+      for (const auto &dimAndTile : dimAndTileMapping) {
+        targetInnerTileSizes[dimAndTile.first] = dimAndTile.second;
+      }
+    } else {
+      // tileSize comes from OpOperand
+      targetInnerTileSizes = unPackOp.getInnerTiles();
+      targetTileSizes = llvm::to_vector(
+          ArrayRef(tileSizes).take_back(targetInnerTileSizes.size()));
+    }
+  }
+
+  // check tileSizes is full on or multiple of `inner_tile_size`
+  for (auto [tile, innerTile] :
+       llvm::zip_equal(targetTileSizes, targetInnerTileSizes)) {
+    if (isEqualConstantIntOrValue(tile, innerTile))
+      continue;
+    FailureOr<int64_t> cstSize = ValueBoundsConstraintSet::computeConstantBound(
+        presburger::BoundType::UB, tile,
+        /*stopCondition=*/nullptr, /*closedUB=*/true);
+    std::optional<int64_t> cstInnerSize = getConstantIntValue(innerTile);
+    if (!failed(cstSize) && cstInnerSize) {
+      if (*cstSize % *cstInnerSize == 0)
+        continue;
```

Comment:
As the name of this filter(`exactTilingOnPackUnPackFilter`) said, uneven tiling/generic case is not expected so far because it will involve more complex [imperfect tiling case](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Tensor/IR/TensorTilingInterfaceImpl.cpp#L396).